internal/blockchain/ethereum/ethereum.go

```diff
@@ -930,121 +920,152 @@ func (e *Ethereum) GenerateFFI(ctx context.Context, generationRequest *core.FFIG
 	ffi := &core.FFI{
 		Namespace:   ns,
 		Name:        name,
 		Version:     version,
 		Description: description,
-		Methods:     []*core.FFIMethod{},
-		Events:      []*core.FFIEvent{},
-	}
-
-	for _, element := range abi {
-		switch element.Type {
-		case "event":
-			event := &core.FFIEvent{
-				FFIEventDefinition: core.FFIEventDefinition{
-					Name:   element.Name,
-					Params: e.convertABIArgumentsToFFI(element.Inputs),
-				},
-			}
-			ffi.Events = append(ffi.Events, event)
-		case "function":
-			method := &core.FFIMethod{
-				Name:    element.Name,
-				Params:  e.convertABIArgumentsToFFI(element.Inputs),
-				Returns: e.convertABIArgumentsToFFI(element.Outputs),
-			}
-			ffi.Methods = append(ffi.Methods, method)
-		}
+		Methods:     make([]*core.FFIMethod, len(abi.Functions())),
+		Events:      make([]*core.FFIEvent, len(abi.Events())),
+	}
+	i := 0
+	for _, f := range abi.Functions() {
+		ffi.Methods[i] = e.convertABIFunctionToFFIMethod(ctx, f)
+		i++
+	}
+	i = 0
+	for _, f := range abi.Events() {
+		ffi.Events[i] = e.convertABIEventToFFIEvent(ctx, f)
+		i++
 	}
 	return ffi
 }
 
-func (e *Ethereum) convertABIArgumentsToFFI(args []ABIArgumentMarshaling) core.FFIParams {
-	ffiParams := core.FFIParams{}
-	for _, arg := range args {
+func (e *Ethereum) convertABIFunctionToFFIMethod(ctx context.Context, abiFunction *abi.Entry) *core.FFIMethod {
+	params := make([]*core.FFIParam, len(abiFunction.Inputs))
+	returns := make([]*core.FFIParam, len(abiFunction.Outputs))
+	for i, input := range abiFunction.Inputs {
+		schema := e.getSchemaForABIInput(ctx, input)
+		param := &core.FFIParam{
+			Name:   input.Name,
+			Schema: fftypes.JSONAnyPtr(schema.ToJSON()),
+		}
+		params[i] = param
+	}
+	for i, input := range abiFunction.Outputs {
+		schema := e.getSchemaForABIInput(ctx, input)
+		param := &core.FFIParam{
+			Name:   input.Name,
+			Schema: fftypes.JSONAnyPtr(schema.ToJSON()),
+		}
+		returns[i] = param
+	}
+	return &core.FFIMethod{
+		Name:    abiFunction.Name,
+		Params:  params,
+		Returns: returns,
+	}
+}
+
+func (e *Ethereum) convertABIEventToFFIEvent(ctx context.Context, abiEvent *abi.Entry) *core.FFIEvent {
+	params := make([]*core.FFIParam, len(abiEvent.Inputs))
+	for i, input := range abiEvent.Inputs {
+		schema := e.getSchemaForABIInput(ctx, input)
 		param := &core.FFIParam{
-			Name: arg.Name,
+			Name:   input.Name,
+			Schema: fftypes.JSONAnyPtr(schema.ToJSON()),
 		}
-		s := e.getSchema(arg)
-		param.Schema = fftypes.JSONAnyPtr(s.ToJSON())
-		ffiParams = append(ffiParams, param)
+		params[i] = param
+	}
+	return &core.FFIEvent{
+		FFIEventDefinition: core.FFIEventDefinition{
+			Name:   abiEvent.Name,
+			Params: params,
+		},
 	}
-	return ffiParams
 }
 
-func (e *Ethereum) getSchema(arg ABIArgumentMarshaling) *Schema {
-	s := &Schema{
-		Type: e.getFFIType(arg.Type),
+func (e *Ethereum) getSchemaForABIInput(ctx context.Context, input *abi.Parameter) *Schema {
```

Comment:
This function still seems to have a bunch of `[]` related processing, and I could not convince myself that something like `(bytes32[5][][10],fixed128x18)` would be handled completely.

None of this should be required with the new API available, it should be possible to simply recurse the parsed component type tree structure:
https://pkg.go.dev/github.com/hyperledger/firefly-signer/pkg/abi#ParameterArray.TypeComponentTreeCtx